### PR TITLE
Chunk Error Fix

### LINKED ISSUE #11847: Add support for traditional websphere property: com.ibm.ws.webcontainer.suppresslastzerobytepackage
Long time ago, in a land far away (or traditional websphere in 2009 or so) this property was added, com.ibm.ws.webcontainer.suppresslastzerobytepackage, due to APAR PK82794.  The code that added the functionality for this custom property did not make the journey from traditional websphere to liberty, and now we have an internal customer that has come across a need for this again,  so we will add it. 

APAR link is: https://www.ibm.com/support/pages/apar/PK82794

### PR DESCRIPTION
**Pull request by FlynnIBM**
Fixes  #11847

### CODE REVIEW

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java [hunk 1]
```diff
@@ -2853,7 +2854,19 @@ final protected void sendFullOutgoing(WsByteBuffer[] wsbb, HttpBaseMessageImpl m
                 }
 
             } else if (msg.isChunkedEncodingSet()) {
-                createEndOfBodyChunk();
+                HttpInboundServiceContextImpl localHisc = null;
+                if (this instanceof HttpInboundServiceContextImpl) {
+                    localHisc = (HttpInboundServiceContextImpl) this;
+                }
+                if (localHisc != null && localHisc instanceof HttpInboundServiceContextImpl
```
Comment (tspewak) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java:2861]: Is the  localHisc instanceof HttpInboundServiceContextImpl   check needed? It was checked above before setting localHisc

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java [hunk 2]
```diff
@@ -2892,8 +2905,18 @@ final protected VirtualConnection sendFullOutgoing(WsByteBuffer[] wsbb, HttpBase
             if (this instanceof HttpInboundServiceContextImpl) {
                 hisc = (HttpInboundServiceContextImpl) this;
             }
-            if (hisc != null && !(hisc.getLink() instanceof H2HttpInboundLinkWrap)) {
+            if (hisc != null && hisc instanceof HttpInboundServiceContextImpl) {
```
Comment (tspewak) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java:2908]: Is && hisc instanceof HttpInboundServiceContextImpl  needed?  it was checked above before setting hisc

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebApp.java [hunk 3]
```diff
@@ -4095,8 +4099,21 @@ public void sendError(HttpServletRequest req, HttpServletResponse res, ServletEr
...
-            reqState.setAttribute("com.ibm.ws.webcontainer.suppresslastzerobytepackage", "true");
+        if (WCCustomProperties.SUPPRESS_LAST_ZERO_BYTE_PACKAGE) {
+           // reqState.setAttribute("com.ibm.ws.webcontainer.suppresslastzerobytepackage", "true");
+            if (req instanceof SRTServletRequest) {
+                SRTServletRequest srtReq = (SRTServletRequest) req;
+                IRequestExtended iReq = (IRequestExtended)srtReq.getIRequest();
+                if (iReq != null) {
+                    HttpInboundConnection httpInboundConnection = iReq.getHttpInboundConnection();
+                    HttpInboundConnectionExtended extendedConnection= (HttpInboundConnectionExtended) httpInboundConnection;
+                    HttpDispatcherLink dispatcherLink=(HttpDispatcherLink) extendedConnection.getHttpDispatcherLink();
```
Comment (tspewak) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/webapp/WebApp.java:4110]: Could extendedConnection ever be null?

**Comment by FlynnIBM** [comment]
#build

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java [hunk 4]
```diff
@@ -1260,4 +1260,14 @@ public boolean useForwardedHeaders() {
         return false;
     }
 
+    /**
+     * @param b
```
Comment (ginnick) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:1264]: Description ?

**Comment by FlynnIBM** [comment]
#build

**Comment by FlynnIBM** [comment]
#run-libby-bot

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java [hunk 5]
```diff
@@ -2892,8 +2903,16 @@ final protected VirtualConnection sendFullOutgoing(WsByteBuffer[] wsbb, HttpBase
             if (this instanceof HttpInboundServiceContextImpl) {
                 hisc = (HttpInboundServiceContextImpl) this;
             }
-            if (hisc != null && !(hisc.getLink() instanceof H2HttpInboundLinkWrap)) {
-                createEndOfBodyChunk();
+            if (hisc != null) {
```
Comment (mrsaldana) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java:2906]: Originally this was checking that the link wasn't an H2HttpInboundLinkWrap, and not doing it now would seem to be a change in behavior. Additionally, by not checking the instanceof of the hisc, this conditional could potentially do a hisc.getSupress0ByteChunk() on an object that isn't an HttpInboundServiceContextImpl, which could lead to exceptions during the runtime. I recommend only doing the new code when the hisc is proven to be HttpInboundServiceContextImpl.

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java [hunk 6]
```diff
@@ -1260,4 +1260,14 @@ public boolean useForwardedHeaders() {
         return false;
     }
 
+    /**
+     * Calls function to set the supress 0 byte chunk flag.
+     */
+    public void setSuppressZeroByteChunk(boolean suppress0ByteChunk) {
+        if (this.isc != null) {
```
Comment (mrsaldana) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:1267]: Might be useful to log if this method isn't able to set the flag due to the isc being null. Nothing code breaking, just QoL change for easier debugging.

**Comment by FlynnIBM** [comment]
#run-libby-bot

**Pull request merged**